### ISSUE TITLE
Top-level API for energy evaluation/analysis

### ISSUE THREAD

**Issue by wodeni** [penrose/penrose#511]
Our top-level API does not allow clients to evaluate constraints/objectives in custom ways. Here are some possible use cases:

* Evaluate the current energy value of a `State`:
```ts
const evalEnergy = (s: State): number;
```
* Evaluate a list of constraints/objectives: this will be useful if a user want to apply a subset of constrs/objs on a `State`
```ts
const evalFns = (fn: Fn[], s: State): number[]; // TODO: constraint/objective eval might be different?
```

Use case:

```ts
const s: State = compileTrio(dsl, sub, sty);
// get the overall energy of the state
const eng: number = evalEnergy(s); 
// get a list of `contains` instances
const containFns: Fn[] = s.constrFns.filter(c => c.fname === "contains");
// only eval `contains` in `s` and ignore all other constraints
const containsEngs: number[] = evalFns(containFns, s); 
// assuming there are more constraints in `s`, the (squared?) sum of `containsEngs` should be smaller than `eng`
```

**Comment by wodeni** [comment]
Other functions per discussion in our team meeting (related: #528)

```ts
// return the gradients for the state, cached if `s` is already initialized and run `prepareState` if not.
evalGradients(s: State): number[]; 
// 
```

**Comment by hypotext** [comment]
Resolved in #512 #527
